### PR TITLE
[NATIVE] Lay groundwork for basic errors

### PR DESCRIPTION
**Pull request by openorclose**
# Features:

- [x] Native!
- [x] Builtins work! 
- [x] Can store constants and letiables (repl works)
- [x] Proper tail calls (Iterative processes are iterative)
- [ ] Enforcing of strict types (`'' * ''` still works)
- [x] Error handling (most errors are now 1-1 with interpreter)
- [ ] stringifying functions (shows the transpiled code currently) 

# How to use: 

Change 

```typescript
const DEFAULT_OPTIONS: IOptions = {
  scheduler: 'async',
  steps: 1000,
  isNativeRunnable: false //change to true
}
```
in `src/index.ts` 

# Please help try and break it!

You can assume member access `obj['string']` is not going to work once strict types are enforced. Find some other way to reach eval, or to break the transpiler. In particular, managing to access the global variable $$NATIVE_STORAGE in Source would break it, as all the functions inside can be switched out.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -40,14 +44,55 @@ export function runInContext(
...
         return Promise.resolve({
           status: 'finished',
-          value: sandboxedEval(generate(transpile(program, context.contextId)))
+          value: sandboxedEval(transpiled)
         } as Result)
       } catch (error) {
-        context.errors.push(new ExceptionError(error, UNKNOWN_LOCATION))
-        return Promise.resolve({ status: 'error' } as Result)
+        if (error instanceof RuntimeSourceError) {
+          context.errors.push(error)
```
Comment (thomastanck) [inline src/index.ts]: Just checking, do these errors come from stuff like type checks being injected into the code? I'm not sure what triggers this line.

File: src/index.ts [hunk 2]
```diff
@@ -40,14 +44,55 @@ export function runInContext(
...
+        if (
+          match === null ||
+          sourceMapJson === undefined ||
+          lastStatementSourceMapJson === undefined
+        ) {
+          context.errors.push(new ExceptionError(error, UNKNOWN_LOCATION))
+          return resolvedErrorPromise
+        }
+        const line = Number(match![1])
+        const column = Number(match![2])
```
Comment (thomastanck) [inline src/index.ts]: Seems like you're parsing the stacktrace, did you consider using a library? A quick google search yields StackTrace.JS, but I'm not sure if overkill.

File: src/transpiler.ts [hunk 3]
```diff
@@ -249,51 +327,67 @@ function getStatementsToAppend(program: es.Program): es.Statement[] {
 
 function splitLastStatementIntoStorageOfResultAndAccessorPair(
   lastStatement: es.Statement
-): es.Statement[] {
+): [es.Statement, es.Statement, {}?] {
```
Comment (thomastanck) [inline src/transpiler.ts]: Same as above, I didn't notice this before but consider using an object instead. Object literal shorthands and destructuring make it quite easy to return and assign these. Objects are more "type safe" compared to arrays as there are names going along with each field.

File: src/utils/rttc.ts [hunk 4]
```diff
@@ -34,18 +34,13 @@ const typeOf = (v: Value) => {
   }
 }
 
-const isNumber = (v: Value) => typeOf(v) === 'number'
-const isString = (v: Value) => typeOf(v) === 'string'
-const isBool = (v: Value) => typeOf(v) === 'boolean'
-const isObject = (v: Value) => typeOf(v) === 'object'
-const isArray = (v: Value) => typeOf(v) === 'array'
+export const isNumber = (v: Value) => typeOf(v) === 'number'
```
Comment (thomastanck) [inline src/utils/rttc.ts]: Reason for exporting all of these?

File: src/index.ts (hunk 1 above)
Comment (openorclose) [reply src/index.ts]: Yes, look at operators.ts, there it throws an InvalidNumberOfArguments RuntimeSourceError.

File: src/index.ts [hunk 5]
```diff
@@ -40,14 +44,55 @@ export function runInContext(
...
     if (theOptions.isNativeRunnable) {
+      let transpiled
+      let sourceMapJson
+      let lastStatementSourceMapJson
       try {
+        const temp = transpile(program, context.contextId)
+        // some issues with formatting and semicolons and tslint so
+        transpiled = temp[0]
+        sourceMapJson = temp[1]
+        lastStatementSourceMapJson = temp[2]
```
[the remark this answers is not in the document]
Comment (openorclose) [reply src/index.ts]: I'll have to work on the destructuring part, ts lint and prettier don't seem to play nice with

```;({a, b, c}) = obj```

But I do agree that returning them as objects is definitely nicer.

File: src/index.ts (hunk 2 above)
Comment (openorclose) [reply src/index.ts]: It's just parsing a single line, nothing too advanced yet.

File: src/transpiler.ts (hunk 3 above)
Comment (openorclose) [reply src/transpiler.ts]: Agree.

File: src/utils/rttc.ts (hunk 4 above)
Comment (openorclose) [reply src/utils/rttc.ts]: Thought i was going to use them, but turns out I didn't

**Pull request merged**